examples/dataflow-csvio/src/main/java/com/google/example/csvio/SortContextualHeadersAndRows.java

```diff
@@ -0,0 +1,427 @@
+import org.apache.beam.sdk.transforms.Min;
+import org.apache.beam.sdk.transforms.PTransform;
+import org.apache.beam.sdk.transforms.ParDo;
+import org.apache.beam.sdk.transforms.SimpleFunction;
+import org.apache.beam.sdk.transforms.View;
+import org.apache.beam.sdk.values.KV;
+import org.apache.beam.sdk.values.PCollection;
+import org.apache.beam.sdk.values.PCollectionTuple;
+import org.apache.beam.sdk.values.PCollectionView;
+import org.apache.beam.sdk.values.PInput;
+import org.apache.beam.sdk.values.POutput;
+import org.apache.beam.sdk.values.PValue;
+import org.apache.beam.sdk.values.Row;
+import org.apache.beam.sdk.values.TupleTag;
+import org.apache.beam.sdk.values.TupleTagList;
+import org.checkerframework.checker.nullness.qual.NonNull;
+import org.checkerframework.checker.nullness.qual.Nullable;
+
+/**
+ * A {@link PTransform} responsible for bundling the raw {@link ContextualTextIO#read()} {@link Row}
+ * {@link PCollection} output into the corresponding CSV headers and rows.
+ */
+@AutoValue
+abstract class SortContextualHeadersAndRows
+    extends PTransform<PCollection<Row>, SortContextualHeadersAndRows.Result> {
+
+  private static final String TAG_BASE = SortContextualHeadersAndRows.class.getSimpleName();
+
+  static Builder builder() {
+    return new AutoValue_SortContextualHeadersAndRows.Builder();
+  }
+
+  static final TupleTag<Row> HEADERS = new TupleTag<>() {};
+  static final TupleTag<Row> ROWS = new TupleTag<>() {};
+
+  abstract CSVIOReadConfiguration getConfiguration();
+
+  @Override
+  public void validate(@Nullable PipelineOptions options) {
+    getConfiguration().validate();
+  }
+
+  @Override
+  public Result expand(PCollection<Row> input) {
+    PCollection<Row> nonEmpty =
+        input.apply(TAG_BASE + "/ExcludeEmptyLines", Filter.by(new IsNotEmptyFn()));
+    CSVIOReadConfiguration configuration = getConfiguration();
+    if (configuration.getHeaderMatchRegex() == null && configuration.getHeaderPosition() == null) {
+      return nonEmpty.apply(
+          TAG_BASE + "/" + SortMinimumHeaderAndRows.class.getSimpleName(),
+          new SortMinimumHeaderAndRows());
+    }
+    if (configuration.getHeaderMatchRegex() != null) {
+      return nonEmpty.apply(
+          TAG_BASE + "/" + SortMatchedHeaderAndRows.class.getSimpleName(),
+          new SortMatchedHeaderAndRows(configuration));
+    }
+    return nonEmpty.apply(
+        TAG_BASE + "/" + SortFixedHeaderAndRows.class.getSimpleName(),
+        new SortFixedHeaderAndRows(configuration));
+  }
+
+  @AutoValue.Builder
+  abstract static class Builder {
+
+    public abstract Builder setConfiguration(CSVIOReadConfiguration value);
+
+    public abstract SortContextualHeadersAndRows build();
+  }
+
+  /**
+   * The result of processing the raw {@link ContextualTextIO#read()} {@link Row} {@link
+   * PCollection} output into the corresponding CSV headers and rows.
+   */
+  static class Result implements POutput, PInput {
```

Comment:
How about calling the class name as SortContextualResult to make it more explicit?